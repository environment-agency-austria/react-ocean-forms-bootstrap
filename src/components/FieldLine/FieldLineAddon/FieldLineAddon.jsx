/**
 * Copyright (c) 2018-present, Umweltbundesamt GmbH
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
import React from 'react';
import PropTypes from 'prop-types';
import { InputGroupAddon, InputGroupText } from 'reactstrap';
import { FormText } from 'react-ocean-forms';

/**
 * Displays a prefix / suffix addon for the FieldLine
 */
function FieldLineAddon({ plaintext, content, type }) {
  if (plaintext) return null;
  if (!content) return null;

  let child = null;
  if (typeof content === 'string') {
    child = (
      <InputGroupText>
        <FormText text={content} />
      </InputGroupText>
    );
  } else {
    child = content;
  }

  return (
    <InputGroupAddon addonType={type}>
      {child}
    </InputGroupAddon>
  );
}

FieldLineAddon.defaultProps = {
  content: null,
};

FieldLineAddon.propTypes = {
  type: PropTypes.string.isRequired,
  content: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.node,
    PropTypes.func,
  ]),
  plaintext: PropTypes.bool.isRequired,
};

export default FieldLineAddon;
