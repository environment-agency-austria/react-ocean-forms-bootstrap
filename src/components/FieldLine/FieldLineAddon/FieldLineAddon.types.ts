export type IFieldLineAddonType = 'prepend' | 'append';

/**
 * Props for the FieldLineAddon component
 */
export interface IFieldLineAddonProps {
  /**
   * Addon type
   */
  type: IFieldLineAddonType;
  /**
   * Form plaintext mode
   */
  plaintext: boolean;
  /**
   * Addon content / child
   */
  content?: React.ReactNode;
}
