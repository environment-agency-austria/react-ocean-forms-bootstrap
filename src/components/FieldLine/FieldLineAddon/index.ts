export * from './FieldLineAddon';
export * from './FieldLineAddon.types';
