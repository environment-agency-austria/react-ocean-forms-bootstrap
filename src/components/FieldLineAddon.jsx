/**
 * Copyright (c) 2018-present, Umweltbundesamt GmbH
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
import React from 'react';
import PropTypes from 'prop-types';
import { InputGroupAddon } from 'reactstrap';
import { fieldMetaShape } from 'react-ocean-forms';

/**
 * Displays a prefix / suffix addon for the FieldLine
 */
function FieldLineAddon({ meta, content, type }) {
  if (meta.plaintext) return null;
  if (!content) return null;

  let child = null;
  if (typeof content === 'string') {
    child = meta.stringFormatter(content);
  } else {
    child = content;
  }

  return (
    <InputGroupAddon addonType={type}>
      {child}
    </InputGroupAddon>
  );
}

FieldLineAddon.defaultProps = {
  content: null,
};

FieldLineAddon.propTypes = {
  type: PropTypes.string.isRequired,
  content: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.node,
    PropTypes.func,
  ]),
  meta: fieldMetaShape.isRequired,
};

export default FieldLineAddon;
