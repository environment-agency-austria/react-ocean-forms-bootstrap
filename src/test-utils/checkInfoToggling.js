/**
 * Re-usable test suite to check the info visibility
 * toggling between InfoAlert and InfoAddonButton
 */
const checkInfoToggling = (wrapper) => {
  describe('info visible toggling', () => {
    const getVisibility = () => wrapper.find('InfoAlert').prop('visible');

    const toggleCases = [
      ['InfoAddonButton', 'onClick'],
      ['InfoAlert', 'onClose'],
    ];

    describe.each(toggleCases)('%s click', (element, prop) => {
      let visibility = getVisibility();

      it(`should toggle the info visibility on ${element} click`, () => {
        for (let i = 0; i < 2; i += 1) {
          wrapper.find(element).prop(prop)();
          const newVisibility = getVisibility();
          expect(newVisibility).toBe(!visibility);
          visibility = newVisibility;
        }
      });
    });
  });
};

export default checkInfoToggling;
