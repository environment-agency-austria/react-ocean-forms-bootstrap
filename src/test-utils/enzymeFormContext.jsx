/**
 * Copyright (c) 2018-present, Umweltbundesamt GmbH
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import createMockFormatter from './createMockFormatter';

/**
 * Components inside the form module require access to the form context.
 * This is not available when mounting single components in Enzyme.
 * These helper functions aim to address that and wrap a valid,
 * user definable form context around them.
 */

/**
 * Creates a form context
 */
export const createMockFormContext = registerCallback => ({
  fieldPrefix: null,

  registerField: jest.fn().mockImplementation((name, state) => registerCallback(name, state)),
  unregisterField: jest.fn(),
  notifyFieldEvent: jest.fn(),

  registerListener: jest.fn().mockImplementation((name, state) => registerCallback(name, state)),
  unregisterListener: jest.fn(),

  getFieldState: jest.fn(),
  getValues: jest.fn(),

  stringFormatter: createMockFormatter(),

  busy: false,
  disabled: false,
  plaintext: false,

  asyncValidateOnChange: false,
  asyncValidationWait: 400,
  defaultValues: {},
});

/**
 * Creates a mock field meta
 */
export const createMockFieldMeta = () => ({
  valid: true,
  error: null,
  isValidating: false,
  touched: false,
  stringFormatter: createMockFormatter(),
  plaintext: false,
});

/**
 * Creates a mock field shape
 */
export const createMockField = () => ({
  id: 'mock-field',
  name: 'mock-field',
  value: 'mock-value',
  disabled: false,
  onChange: jest.fn(),
  onBlur: jest.fn(),
});
