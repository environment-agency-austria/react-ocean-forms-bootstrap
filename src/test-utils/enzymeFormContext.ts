/**
 * Copyright (c) 2018-present, Umweltbundesamt GmbH
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
import { IFieldComponentFieldProps, IFieldComponentMeta, IFieldState, IFormContext, TFormEventListener } from 'react-ocean-forms';

import { createMockFormatter } from './createMockFormatter';

/**
 * Components inside the form module require access to the form context.
 * This is not available when mounting single components in Enzyme.
 * These helper functions aim to address that and wrap a valid,
 * user definable form context around them.
 */

/**
 * Creates a form context
 */
export const createMockFormContext = (registerCallback?: Function): IFormContext => ({
  fieldPrefix: null,

  registerField: jest.fn().mockImplementation((name: string, state: IFieldState): void => {
    if (registerCallback) {
      registerCallback(name, state);
    }
  }),
  unregisterField: jest.fn(),
  notifyFieldEvent: jest.fn(),

  registerListener: jest.fn().mockImplementation((name: string, callback: TFormEventListener): void => {
    if (registerCallback) {
      registerCallback(name, callback);
    }
  }),
  unregisterListener: jest.fn(),

  getFieldState: jest.fn(),
  getValues: jest.fn(),

  stringFormatter: createMockFormatter(),
  submit: jest.fn(),

  busy: false,
  disabled: false,
  plaintext: false,

  asyncValidateOnChange: false,
  asyncValidationWait: 400,
  defaultValues: {},
});

/**
 * Creates a mock field meta
 */
export const createMockFieldMeta = (): IFieldComponentMeta => ({
  valid: true,
  error: null,
  isValidating: false,
  isRequired: false,
  touched: false,
  stringFormatter: createMockFormatter(),
  plaintext: false,
});

/**
 * Creates a mock field shape
 */
export const createMockField = (): IFieldComponentFieldProps => ({
  id: 'mock-field',
  name: 'mock-field',
  value: 'mock-value',
  disabled: false,
  onChange: jest.fn(),
  onBlur: jest.fn(),
});
