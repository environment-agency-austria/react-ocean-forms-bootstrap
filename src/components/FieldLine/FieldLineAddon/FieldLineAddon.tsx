/**
 * Copyright (c) 2018-present, Umweltbundesamt GmbH
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
import React from 'react';

import { FormText } from 'react-ocean-forms';
import { InputGroupAddon, InputGroupText } from 'reactstrap';

import { IFieldLineAddonProps } from './FieldLineAddon.types';

/**
 * Displays a prefix / suffix addon for the FieldLine
 */
export const FieldLineAddon: React.FC<IFieldLineAddonProps> = (props) => {
  const { plaintext, content, type } = props;

  if (plaintext || content === undefined) { return null; }

  let child = null;
  if (typeof content === 'string') {
    child = (
      <InputGroupText>
        <FormText text={content} />
      </InputGroupText>
    );
  } else {
    child = content;
  }

  return (
    <InputGroupAddon addonType={type}>
      {child}
    </InputGroupAddon>
  );
};
