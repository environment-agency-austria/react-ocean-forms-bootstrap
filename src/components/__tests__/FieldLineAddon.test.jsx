import React from 'react';
import { shallow } from 'enzyme';

import FieldLineAddon from '../FieldLineAddon';
import { createMockFieldMeta } from '../../test-utils/enzymeFormContext';

describe('<FieldLineAddon />', () => {
  const setup = (plaintext, content, type = 'prepend') => {
    const meta = createMockFieldMeta();
    meta.plaintext = plaintext;

    const wrapper = shallow((
      <FieldLineAddon
        meta={meta}
        content={content}
        type={type}
      />
    ));

    return { wrapper, meta };
  };

  const checkAddonRender = (wrapper, type) => {
    it('should render an InputGroupAddon', () => {
      expect(wrapper.exists('InputGroupAddon')).toBeTruthy();
    });

    it('should pass the correct addon type', () => {
      expect(wrapper.find('InputGroupAddon').prop('addonType')).toBe(type);
    });
  };

  it('should render nothing if meta.plaintext is active', () => {
    const { wrapper } = setup(true);
    expect(wrapper.exists('InputGroupAddon')).toBeFalsy();
  });

  it('should render nothing if there is no content', () => {
    const { wrapper } = setup(false, null);
    expect(wrapper.exists('InputGroupAddon')).toBeFalsy();
  });

  describe('text content', () => {
    const MOCK_CONTENT = 'mock-string';
    const MOCK_TYPE = 'prepend';
    const { wrapper, meta } = setup(false, MOCK_CONTENT, MOCK_TYPE);

    it('should pass the text content through meta.stringFormatter', () => {
      expect(meta.stringFormatter).toHaveBeenCalledWith(MOCK_CONTENT);
    });

    describe('render', () => {
      checkAddonRender(wrapper, MOCK_TYPE);

      it('should render correctly', () => {
        expect(wrapper).toMatchSnapshot();
      });
    });
  });

  describe('custom content', () => {
    const MOCK_CONTENT = <div id="mock-content" />;
    const MOCK_TYPE = 'append';
    const { wrapper, meta } = setup(false, MOCK_CONTENT, MOCK_TYPE);

    it('should not pass the content through meta.stringFormatter', () => {
      expect(meta.stringFormatter).not.toHaveBeenCalledWith(MOCK_CONTENT);
    });

    describe('render', () => {
      checkAddonRender(wrapper, MOCK_TYPE);

      it('should render correctly', () => {
        expect(wrapper).toMatchSnapshot();
      });
    });
  });
});
