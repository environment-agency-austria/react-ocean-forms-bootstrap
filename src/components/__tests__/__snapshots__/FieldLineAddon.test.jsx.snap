// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FieldLineAddon /> custom content render should render correctly 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <FieldLineAddon
    content={
      <div
        id="mock-content"
      />
    }
    plaintext={false}
    type="append"
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "addonType": "append",
      "children": <div
        id="mock-content"
      />,
      "tag": "div",
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "id": "mock-content",
      },
      "ref": null,
      "rendered": null,
      "type": "div",
    },
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "addonType": "append",
        "children": <div
          id="mock-content"
        />,
        "tag": "div",
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "id": "mock-content",
        },
        "ref": null,
        "rendered": null,
        "type": "div",
      },
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getSnapshotBeforeUpdate": true,
        },
      },
    },
    "attachTo": undefined,
    "hydrateIn": undefined,
  },
}
`;

exports[`<FieldLineAddon /> text content render should render correctly 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <FieldLineAddon
    content="mock-string"
    plaintext={false}
    type="prepend"
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "addonType": "prepend",
      "children": <FormComponent(FormText)
        text="mock-string"
      />,
      "tag": "div",
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "text": "mock-string",
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "addonType": "prepend",
        "children": <FormComponent(FormText)
          text="mock-string"
        />,
        "tag": "div",
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "text": "mock-string",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getSnapshotBeforeUpdate": true,
        },
      },
    },
    "attachTo": undefined,
    "hydrateIn": undefined,
  },
}
`;
